src/AggregateFunctions/AggregateFunctionMinMaxAny.h

```diff
@@ -474,78 +505,130 @@ struct SingleValueDataString //-V730
+            capacity = static_cast<UInt32>(rounded_capacity);
 
-                if (size > 0)
-                    buf.readStrict(small_data, size);
-            }
-            else
-            {
-                if (capacity < rhs_size)
-                {
-                    capacity = static_cast<Int32>(roundUpToPowerOfTwoOrZero(rhs_size));
-                    /// It might happen if the size was too big and the rounded value does not fit a size_t
-                    if (unlikely(capacity < rhs_size))
-                        throw Exception(ErrorCodes::TOO_LARGE_STRING_SIZE, "String size is too big ({})", rhs_size);
+            /// Don't free large_data here.
+            large_data = arena->alloc(capacity);
+        }
+    }
 
-                    /// Don't free large_data here.
-                    large_data = arena->alloc(capacity);
-                }
+    void read(ReadBuffer & buf, const ISerialization & /*serialization*/, Arena * arena)
+    {
+        /// For serialization we use signed Int32 (for historical reasons), -1 means "no value"
+        Int32 rhs_size_signed;
+        readBinary(rhs_size_signed, buf);
 
-                size = rhs_size;
-                buf.readStrict(large_data, size);
-            }
+        if (rhs_size_signed < 0)
+        {
+            /// Don't free large_data here.
+            size = 0;
+            return;
         }
-        else
+
+        UInt32 rhs_size = rhs_size_signed;
+        if (rhs_size <= MAX_SMALL_STRING_SIZE)
         {
             /// Don't free large_data here.
             size = rhs_size;
+            buf.readStrict(small_data, size);
+        }
+        else
+        {
+            /// Reserve one byte more for null-character
+            Int32 rhs_size_to_reserve = rhs_size + 1;
+            allocateLargeDataIfNeeded(rhs_size_to_reserve, arena);
+            size = rhs_size;
+            buf.readStrict(large_data, size);
+        }
+
+        /// Check if the string we read is null-terminated (getDataMutable does not have the assertion)
+        if (0 < size && getDataMutable()[size - 1] == '\0')
+            return;
+
+        /// It's not null-terminated, but it must be (for historical reasons). There are two variants:
+        /// - The value was serialized by one of the incompatible versions of ClickHouse. We had some range of versions
+        ///   that used to serialize SingleValueDataString without terminating '\0'. Let's just append it.
+        /// - An attacker sent crafted data. Sanitize it and append '\0'.
+        /// In all other cases the string must be already null-terminated.
+
+        /// NOTE We cannot add '\0' unconditionally, because it will be duplicated.
+        /// NOTE It's possible that a string that actually ends with '\0' was written by one of the incompatible versions.
+        ///      Unfortunately, we cannot distinguish it from normal string written by normal version.
+        ///      So such strings will be trimmed.
```

Comment:
Not sure that simple comment in the code is enough for this case, even though it is really uncommon case.
Apart from this LGTM.

Reply:
Yep, I agree, of course we will add a note about consequences of the issue to the changelog.

Reply:
Just an idea, maybe it worth printing backward incompatible changes after upgrade (if any) during server/client first start, though it will require built changelog into clickhouse

Reply:
I'm not sure, it's not clear what is "first start". It's not really useful to just print warnings to server log, because almost nobody will notice these warnings. It makes sense to print them when client connects to the server at the first time after upgrade, but what if multiple clients connect to the server. And it's easy miss a message that is shown only once. And usually it's too late to warn about incompatibility _after_ upgrade.